docs/src/architecture/backend_architecture.md

```diff
@@ -0,0 +1,220 @@
+
+Having a p2p solution is not possible due to potential cheats and due to connectivity problems between cellphones.
+The server is responsible of coordinating state between clients and banning cheaters.
+
+The architecture of the server should be as simple as possible. 
+
+- 1 process per player
+- 1 game_state process
+- N+1 processes in total with N being the number of players
+
+## Round
+1. Clients send messages to the `game_state` process
+2. Game state processes all messages in order
+3. The `tick`, which is the clock of the game, is updated and the game state process sends a small delta update to all the players with the changes that were realized during the tick.
+
+> Most of the work is done during 3 where collision is detected, powers get consumed, creates get spawned or assigned to a client, damage and death are resolved.
+4. Go to 1.
+
+## API
+We aim for the API provided to clients during a game to be as simple as possible:
+
+- move(joystick_values): Allows a player to move through the board by providing joystick values corresponding to the direction of movement.
+- basic_attack(joystick_values): Takes joystick values to determine the direction of the attack.
+- skill(skill_name, joystick_values): Accepts joystick values for skill direction along with the skill name.
+- refresh: Used to request the entire game state.
+
+Additionally, the server can communicate various messages to the clients:
+
+- game_start: Indicates the start of the game.
+- game_end: Signals the end of the game.
+- state: Represents the differences between two states.
+- game_state: Contains all the information about the game.
+- ping: Used to inquire about the server's ping.
+
+## Matchmaking
+
+For matchmaking to work, players should be able to:
+
+- Join an existing lobby.
+- Create a lobby if no other exists.
+- Start the game once the lobby is full.
+
+All matchmaking sessions are spawned by a `DynamicSupervisor` called `MatchingSupervisor`. When a player creates a lobby, this supervisor starts a new child `MatchingSession` process. This is the process that will both handle the logic and hold all of the state for that lobby. Right now, that's just a list of the current players in the lobby, along with the ability to add/remove players and start the game.
+
+Inside a lobby, when the game starts, the lobby process uses [Phoenix PubSub](https://hexdocs.pm/phoenix_pubsub/Phoenix.PubSub.html) to broadcast a message saying that the game has started; then it terminates. We'll go into it later, but the idea is that the *gameplay* part of the server will pick up this message and start the game from there.
+
+Every lobby has a unique `ID`, which we call the `session_id` throughout the code. This id is simply the erlang `PID` of the lobby process with an encoding on top to make it human readable. This way players can pass it around to their friends to join their lobbies.
```

Comment:
this has changed since we wrote it. 

- players can only *join a game*
- the game starts automatically when we reach 10 players or we reach a time limit (in that case the game is filled with bots)
- there is a single process taking care of the matchmaking (the `MatchingCoordinator`)